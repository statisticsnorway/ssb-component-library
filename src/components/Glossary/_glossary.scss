.ssb-glossary {
  @include focus-marker;
  @include reset-button;
  cursor: pointer;
  display: inline-block;
  position: relative;

  .glossary-text-wrap {
    border-bottom: 1px dotted $ssb-green-4;
    display: inline-block;
    line-height: inherit;
  }

  .glossary-logo {
    color: $ssb-green-3;
    margin: 2px;
    transition: color .2s;
    vertical-align: top;
  }

  .glossary-animate-background {
    background: $ssb-green-2;
    bottom: 0;
    height: 0;
    left: 0;
    position: absolute;
    transition: height .2s;
    width: 100%;
    z-index: -1;
  }

  &:hover, &:focus {

    .glossary-animate-background {
      height: 100%;
    }

    .glossary-logo {
      color: $ssb-dark-6;
    }
  }

  .glossary-popup {
    @include open-sans;
    animation: fadein .2s;
    display: none;
    font-size: 14px;
    left: 50%;
    margin-left: -135px;
    opacity: 0;
    position: absolute;
    top: 34px;
    transition: opacity .18s;
    width: 270px;

    &:after {
      border-color: transparent transparent $ssb-dark-5 transparent;
      border-style: solid;
      border-width: 12px;
      content: '';
      left: 50%;
      margin-left: -12px;
      position: absolute;
      top: -22px;
    }

    &.open {
      display: inline-block;
      opacity: 1;
    }

    .content-box {
      background: $ssb-dark-5;
      color: $ssb-white;
      padding: 20px;

      .ssb-glossary-closing {
        align-items: center;
        display: flex;
        justify-content: center;
        margin-top: 10px;
        user-select: none;
        width: 100%;

        .icon {
          color: $ssb-dark-5;
          fill: $ssb-green-2;
          font-size: 16px;
        }

        span {
          line-height: 1.43;
          margin-left: 2px;
        }
      }

      .info-text {
        color: $ssb-white;
        font-size: 14px;
      }
    }
  }
}
