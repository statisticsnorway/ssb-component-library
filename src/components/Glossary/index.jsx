import React, { useEffect, useState, useRef } from 'react';
import PropTypes from 'prop-types';
import { BookOpen, XCircle } from 'react-feather';

const Glossary = ({ explanation, children, closeText }) => {
	const node = useRef();
	const infoContainer = useRef();
	const [open, setOpen] = useState(false);

	const handleClickOutside = e => {
		if (node.current.contains(e.target)) {
			return;
		}
		setOpen(false);
	};

	useEffect(() => {
		if (open) {
			document.addEventListener('mousedown', handleClickOutside);
		} else {
			document.removeEventListener('mousedown', handleClickOutside);
		}

		return () => {
			document.removeEventListener('mousedown', handleClickOutside);
		};
	}, [open]);

	return (
		<button ref={node} onClick={() => setOpen(!open)} className="ssb-glossary">
			<div className="glossary-text-wrap">{children}</div>
			<BookOpen size={12} className="glossary-logo" />
			<div className="glossary-animate-background" />
			{open && (
				<div className={`glossary-popup${open ? ' open' : ''}`} ref={infoContainer}>
					<div className="content-box">
						<span className="info-text">{explanation}</span>
						<div className="ssb-glossary-closing">
							<XCircle size={16} className="icon" /><span>{closeText}</span>
						</div>
					</div>
				</div>
			)}

		</button>
	);
};

Glossary.defaultProps = {
	closeText: 'Lukk',
};

Glossary.propTypes = {
	children: PropTypes.node,
	closeText: PropTypes.string,
	explanation: PropTypes.string.isRequired,
};

export default Glossary;
