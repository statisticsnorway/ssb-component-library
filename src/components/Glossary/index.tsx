import React, { useEffect, useState, useRef, useCallback, type ReactNode } from 'react'
import { BookOpen, XCircle } from 'react-feather'
import classNames from '../../utils/utils'

interface GlossaryProps {
  children?: ReactNode
  className?: string
  closeText?: string
  explanation: string
  ariaLabel?: string
}

const Glossary: React.FC<GlossaryProps> = ({ explanation, children, className, closeText = 'Lukk', ariaLabel }) => {
  const node = useRef<HTMLSpanElement>(null)
  const [open, setOpen] = useState(false)

  const handleClickOutside = (e: MouseEvent) => {
    if (node.current?.contains(e.target as Node)) {
      return
    }
    setOpen(false)
  }

  const handleCloseButton = () => {
    setOpen(false)
    node.current?.focus()
  }

  const handleKeyDownClose = (e: React.KeyboardEvent<HTMLButtonElement>) => {
    if (e.key === 'Enter' || e.key === ' ') {
      e.preventDefault()
      setOpen(false)
      node.current?.focus()
    }
  }

  const escKeyListener = useCallback((e: KeyboardEvent) => {
    if (e.key === 'Escape') {
      setOpen(false)
      node.current?.focus()
    }
  }, [])

  useEffect(() => {
    if (open) {
      document.addEventListener('mousedown', handleClickOutside)
      document.addEventListener('keydown', escKeyListener, false)
    }

    return () => {
      document.removeEventListener('mousedown', handleClickOutside)
      document.removeEventListener('keydown', escKeyListener, false)
    }
  }, [open])

  return (
    <span className={classNames('ssb-glossary', className)}>
      <button
        onClick={() => setOpen(!open)}
        className='glossary-button'
        aria-label={ariaLabel || (typeof children === 'string' ? children : undefined)}
        aria-expanded={open ? 'true' : 'false'}
      >
        <span className='glossary-text-wrap'>{children}</span>
        <BookOpen size={12} className='glossary-logo' aria-hidden='true' />
      </button>
      <span className={classNames('glossary-popup', open && 'open')} ref={node}>
        <span className='content-box'>
          <span className='info-text'>{explanation}</span>
          <span className='glossary-closing'>
            <button
              onClick={() => {
                handleCloseButton()
              }}
              onKeyDown={(e) => handleKeyDownClose(e)}
            >
              <XCircle size={16} className='icon' aria-hidden='true' />
              <span>{closeText}</span>
            </button>
          </span>
        </span>
      </span>
    </span>
  )
}

export default Glossary
