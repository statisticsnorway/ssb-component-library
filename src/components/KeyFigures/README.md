KeyFigures
========

> A component for showing key figures with logo

### Usage

#### HTML

```html
<div class="ssb-key-figures">
  <div class="kf-icon large"><i /></div>
  <div>
    <div class="ssb-title"><h4>Title</h4></div>
    <div>2019</div>
    <div class="ssb-number large">12 345</div>
    <div class="ssb-title"><h4>husholdninger</h4></div>
  </div>
</div>

<div class="ssb-key-figures">
  <div class="kf-icon small"><i /></div>
  <div>
    <div class="ssb-title"><h4>Title</h4></div>
    <div>2019</div>
    <div class="ssb-number small">12 345</div>
    <div class="ssb-title"><h4>husholdninger</h4></div>
  </div>
</div>
```

#### React

```jsx harmony
<KeyFigures
  icon={<Icon />}
  number="12345"
  numberDescription="husholdninger"
  size="small"
  title="Title"
  year="2019"
/>
```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| icon | node | Renders an icon |
| number | string or number | Large number to be displayed |
| numberDescription | string | Subtitle |
| size | 'small', 'medium' or 'large' | Change size of number and icon |
| title | string | Displays title |
| year | string or number | Displays a year between title and number |
