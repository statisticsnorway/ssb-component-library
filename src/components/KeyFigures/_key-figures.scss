.ssb-key-figures {
  @include roboto;
  column-gap: $global-padding-size * 2;
  display: inline-grid;
  grid-auto-flow: column;

  .kf-icon {
    color: $ssb-dark-5;
    display: inline-flex;

    &.large {
      max-height: 230px;
      max-width: 230px;
      width: 230px;

      & > * {
        height: 100%;
        width: 100%;
      }
    }

    &.medium {
      max-height: 180px;
      max-width: 180px;
      width: 180px;

      & > * {
        height: 100%;
        width: 100%;
      }
    }

    &.small {
      max-height: 130px;
      max-width: 130px;
      width: 130px;

      & > * {
        height: 100%;
        width: 100%;
      }
    }
  }

  .kf-time {
    @include open-sans;
    font-size: 16px;
    margin-top: 6px;
  }

  .kf-title {
    @include roboto;
    font-size: 20px;
    font-weight: bold;
    line-height: normal;
    margin: 0;

    &.subtitle {
      color: $ssb-dark-5;
    }
  }
}
