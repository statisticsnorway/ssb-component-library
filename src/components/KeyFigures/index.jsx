import React from 'react';
import PropTypes from 'prop-types';
import Number from '../Number';
import WordExplanation from '../WordExplanation';

const KeyFigures = ({	icon, number, numberDescription, size, title, time, wordExplanation }) => (
	<div className="ssb-key-figures">
		<div className={`kf-icon ${size}`}>{icon}</div>
		<div>
			{wordExplanation
				? (
					<WordExplanation explanation={wordExplanation}>
						<span className="kf-title">{title}</span>
					</WordExplanation>
				)
				: <span className="kf-title">{title}</span>}
			<div className="kf-time">{time}</div>
			<Number size={size} number={number} />
			<span className="kf-title subtitle">{numberDescription}</span>
		</div>
	</div>
);

KeyFigures.defaultProps = {};

KeyFigures.propTypes = {
	icon: PropTypes.node,
	number: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.number,
	]),
	numberDescription: PropTypes.string,
	size: PropTypes.oneOf(['small', 'medium', 'large']),
	title: PropTypes.string,
	time: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.number,
	]),
	wordExplanation: PropTypes.string,
};

export default KeyFigures;
