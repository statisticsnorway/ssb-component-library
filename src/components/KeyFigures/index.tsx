import React, { type ReactNode } from 'react'
import { ArrowDown, ArrowUp, Minus } from 'react-feather'
import Number from '../Number'
import Glossary from '../Glossary'

interface Changes {
  changeDirection?: 'up' | 'down' | 'same'
  changeText?: string
  changePeriod?: string
  srChangeText?: string
}
interface KeyFiguresProps {
  changes: Changes
  className?: string
  icon: ReactNode
  number: string | number
  numberDescription?: string
  noNumberText?: string
  size: 'small' | 'medium' | 'large'
  title?: string
  time: string | number
  glossary?: string
  greenBox?: boolean
}

const KeyFigures: React.FC<KeyFiguresProps> = ({
  changes = {},
  className,
  icon,
  number,
  numberDescription,
  noNumberText = 'Tall ikke tilgjengelig',
  size,
  title,
  time,
  glossary,
  greenBox,
}) => {
  const { changeDirection, changeText, changePeriod, srChangeText } = changes
  return (
    <div className={`ssb-key-figures ${size}${greenBox ? ' green-box' : ''}${className ? ` ${className}` : ''}`}>
      {icon && <div className={`kf-icon ${size}`}>{icon}</div>}
      <div>
        {glossary && title ? (
          <Glossary explanation={glossary} ariaLabel={title}>
            <span className='kf-title'>{title}</span>
          </Glossary>
        ) : (
          <span className='kf-title'>{title}</span>
        )}
        <div className='kf-time'>{time}</div>
        {number ? (
          <div className='number-section'>
            <Number size={size}>{number}</Number>
            <span className='kf-title subtitle'>{numberDescription}</span>
          </div>
        ) : (
          <span className='no-number'>{noNumberText}</span>
        )}
        {changes && (
          <div className='kf-changes'>
            {changeDirection === 'up' && <ArrowUp className='changes-icon' size={20} aria-hidden='true' />}
            {changeDirection === 'down' && <ArrowDown className='changes-icon' size={20} aria-hidden='true' />}
            {changeDirection === 'same' && <Minus className='changes-icon' size={20} aria-hidden='true' />}
            <span className='changes-text' aria-hidden={srChangeText ? 'true' : 'false'}>
              {changeText}
            </span>
            &nbsp;
            <span className='changes-periode' aria-hidden={srChangeText ? 'true' : 'false'}>
              {changePeriod}
            </span>
            {srChangeText ? <span className='sr-only'>{srChangeText}</span> : null}
          </div>
        )}
      </div>
    </div>
  )
}

export default KeyFigures
