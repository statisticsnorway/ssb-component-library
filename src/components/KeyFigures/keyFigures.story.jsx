import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import { Home } from 'react-feather';
import KeyFigures from './index';

const title = 'Antall husholdninger';

storiesOf('KeyFigures', module).addDecorator(centered)
	.add('Large', () => (
		<div>
			<KeyFigures
				number="789 398"
				title={title}
				numberDescription="husholdninger"
				year="2018"
				size="large"
				icon={<Home />}
			/>
		</div>
	))
	.add('Medium', () => (
		<div>
			<KeyFigures
				number="789 398"
				title={title}
				numberDescription="husholdninger"
				year="2018"
				size="medium"
				icon={<Home />}
			/>
		</div>
	))
	.add('Small', () => (
		<div>
			<KeyFigures
				number="789 398"
				title={title}
				numberDescription="husholdninger"
				year="2018"
				size="small"
				icon={<Home />}
			/>
		</div>
	));
