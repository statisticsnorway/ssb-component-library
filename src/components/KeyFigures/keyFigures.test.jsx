import React from 'react';
import {shallow} from 'enzyme';
import { Home } from 'react-feather';
import KeyFigures from './index';
import Glossary from '../Glossary';

const title = 'Antall husholdninger';

describe('KeyFigures component', () => {
	test('Matches the snapshot', () => {
		const wrapper = shallow(
			<KeyFigures
				number="789 398"
				title={title}
				numberDescription="husholdninger"
				time="2018"
				size="large"
				icon={<Home className="feather-home" />}
				glossary={'Explain something'}
			/>
		);
		expect(wrapper).toMatchSnapshot();
	});

	test('Renders Glossary conditionally', () => {
		const wrapper = shallow(
			<KeyFigures number="789 398" glossary={'Explain something'} />
		);
		expect(wrapper.find(Glossary)).toHaveLength(1);

		const wrapper2 = shallow(
			<KeyFigures number="789 398" />
		);
		expect(wrapper2.find(Glossary)).toHaveLength(0);
	});
	test('Renders time and number', () => {
		const wrapper = shallow(<KeyFigures number="789 398" time="2018" />);
		expect(wrapper.find('.kf-time').render().text()).toBe('2018');
		expect(wrapper.find('Number').render().text()).toBe('789 398');
	});
	test('Render no number text', () => {
		const wrapper = shallow(<KeyFigures title="KeyFigure without number" time="2018" noNumberText="Ingen tall"/>);
		expect(wrapper.find('.no-number').render().text()).toBe('Ingen tall');
	});
	test('Render changes', () => {
		const wrapper = shallow(<KeyFigures
			title="KeyFigure without number"
			time="2018"
				changes={{
				changeDirection: 'down',
				changeText: 'Ned 1 prosentpoeng',
				changePeriod: 'fra året før',
			}}
		/>);
		// console.log(wrapper.debug());
		expect(wrapper.find('.kf-changes').exists()).toEqual(true);
		expect(wrapper.find('ArrowDown').exists()).toEqual(true);
		// expect(wrapper.find('.kf-changes').render().text()).toBe('Ingen tall');
	});
});
