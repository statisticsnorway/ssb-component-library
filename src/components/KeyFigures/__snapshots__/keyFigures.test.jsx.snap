// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyFigures component Matches the snapshot 1`] = `
<div
  className="ssb-key-figures large"
>
  <div
    className="kf-icon large"
  >
    <Home
      className="feather-home"
      color="currentColor"
      size="24"
    />
  </div>
  <div>
    <Glossary
      closeText="Lukk"
      explanation="Explain something"
    >
      <span
        className="kf-title"
      >
        Antall husholdninger
      </span>
    </Glossary>
    <div
      className="kf-time"
    >
      2018
    </div>
    <div
      className="number-section"
    >
      <Number
        size="large"
      >
        789 398
      </Number>
      <span
        className="kf-title subtitle"
      >
        husholdninger
      </span>
    </div>
  </div>
</div>
`;
